src/Avalonia.Controls/Presenters/PanelContainerGenerator.cs

```diff
@@ -0,0 +1,161 @@
+            if (_presenter.ItemsControl is { } itemsControl)
+            {
+                itemsControl.PropertyChanged -= OnItemsControlPropertyChanged;
+                itemsControl.ItemsView.PostCollectionChanged -= OnItemsChanged;
+
+                ClearItemsControlLogicalChildren();
+            }
+
+            _presenter.Panel?.Children.Clear();
+        }
+
+        internal void Refresh() => OnItemsChanged(null, CollectionUtils.ResetEventArgs);
+
+        private void OnItemsControlPropertyChanged(object? sender, AvaloniaPropertyChangedEventArgs e)
+        {
+            if (e.Property == ItemsControl.ItemsProperty)
+            {
+                if (e.OldValue is INotifyCollectionChanged inccOld)
+                    inccOld.CollectionChanged -= OnItemsChanged;
+                OnItemsChanged(null, CollectionUtils.ResetEventArgs);
+                if (e.NewValue is INotifyCollectionChanged inccNew)
+                    inccNew.CollectionChanged += OnItemsChanged;
+            }
+        }
+
+        private void OnItemsChanged(object? sender, NotifyCollectionChangedEventArgs e)
+        {
+            if (_presenter.Panel is null || _presenter.ItemsControl is null)
+                return;
+
+            var itemsControl = _presenter.ItemsControl;
+            var generator = itemsControl.ItemContainerGenerator;
+            var children = _presenter.Panel.Children;
+
+            void Add(int index, IEnumerable items)
+            {
+                var i = index;
+                foreach (var item in items)
+                    InsertContainer(itemsControl, children, item, i++);
+
+                var childCount = children.Count;
+                var delta = i - index;
+
+                for (; i < childCount; ++i)
+                    generator.ItemContainerIndexChanged(children[i], i - delta, i);
+            }
+
+            void Remove(int index, int count)
+            {
+                for (var i = 0; i < count; ++i)
+                {
+                    var c = children[i];
+                    if (!c.IsSet(ItemIsOwnContainerProperty))
+                        itemsControl.RemoveLogicalChild(children[i + index]);
+                }
+
+                children.RemoveRange(index, count);
+
+                var childCount = children.Count;
+
+                for (var i = index; i < childCount; ++i)
+                    generator.ItemContainerIndexChanged(children[i], i + count, i);
+            }
+
+            switch (e.Action)
+            {
+                case NotifyCollectionChangedAction.Add:
+                    Add(e.NewStartingIndex, e.NewItems!);
+                    break;
+                case NotifyCollectionChangedAction.Remove:
+                    Remove(e.OldStartingIndex, e.OldItems!.Count);
+                    break;
+                case NotifyCollectionChangedAction.Replace:
+                    Remove(e.OldStartingIndex, e.OldItems!.Count);
+                    Add(e.NewStartingIndex, e.NewItems!);
```

Comment:
Ideally, I think replace actions should try to reuse the existing container instead of remove and add. Per WPF's source, they do a `IsItemItsOwnContainer` check for this, if it is, you need a new container; otherwise it just swaps the DataContext/Content out.

Reply:
Yeah this could be optimized I guess, though I feel like it's a bit of a niche case. I'm not even aware of a list API which has bulk replacement of items so for most cases where it would show up as a performance problem it's going to be a remove followed by an add anyway.

Reply:
@grokys ReactiveUI DynamicData uses items replacement during sorting IIRC.

Reply:
Ok, that's good to know. Still, I'd like to address this in a separate PR if necessary as there's already enough in this one.